File: android/feature/events/src/main/java/com/creapolis/solennix/feature/events/pdf/ContractPdfGenerator.kt
package com.creapolis.solennix.feature.events.pdf

import android.content.Context
import android.graphics.pdf.PdfDocument
import com.creapolis.solennix.core.model.Client
import com.creapolis.solennix.core.model.DiscountType
import com.creapolis.solennix.core.model.Event
import com.creapolis.solennix.core.model.EventExtra
import com.creapolis.solennix.core.model.EventProduct
import com.creapolis.solennix.core.model.User
import java.io.File
import java.io.FileOutputStream
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

object ContractPdfGenerator {

    private val TOKEN_REGEX = Regex("\\[([^\\[\\]]+)\\]")
    private val DATE_FMT = DateTimeFormatter.ofPattern("d 'de' MMMM 'de' yyyy", Locale("es", "MX"))

    fun generate(
        context: Context,
        event: Event,
        client: Client,
        products: List<EventProduct>,
        extras: List<EventExtra>,
        user: User?
    ): File {
        val document = PdfDocument()
        val manager = PdfPageManager(document)
        manager.startNewPage()
        drawHeader(manager, user)

        val template = user?.contractTemplate?.takeIf { it.isNotBlank() }
        if (template != null) {
            renderFromTemplate(manager, template, event, client, products, user)
        } else {
            renderHardcoded(manager, event, client, products, extras, user)
        }

        val businessName = user?.businessName ?: "Solennix"
        drawSignatureSection(manager, businessName, client.name)
        manager.drawFooter("Solennix - Contrato de Servicios")
        manager.finishDocument()
        val file = File(context.cacheDir, "contrato_${event.id.take(8)}.pdf")
        FileOutputStream(file).use { fos -> document.writeTo(fos) }
        document.close()
        return file
    }

    private fun renderFromTemplate(
        manager: PdfPageManager, template: String,
        event: Event, client: Client, products: List<EventProduct>, user: User?
    ) {
        val tokens = buildTokenMap(event, client, products, user)
        val rendered = TOKEN_REGEX.replace(template) { match ->
            val label = match.groupValues[1].trim()
            tokens[label] ?: tokens[normalize(label)] ?: match.value
        }
        rendered.split("\n").forEach { line ->
            if (line.isBlank()) manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
            else manager.drawMultilineText(line)
        }
        manager.moveDown(PdfConstants.SECTION_SPACING)
    }

    private fun normalize(s: String) = s.trim().lowercase()
        .replace("á","a").replace("é","e").replace("í","i")
        .replace("ó","o").replace("ú","u").replace("ñ","n")

    private fun buildTokenMap(
        event: Event, client: Client, products: List<EventProduct>, user: User?
    ): Map<String, String> {
        val name = user?.name ?: ""
        val biz = user?.businessName?.takeIf { it.isNotBlank() } ?: name
        val t0 = event.startTime?.takeIf { it.isNotBlank() }
        val t1 = event.endTime?.takeIf { it.isNotBlank() }
        val range = if (t0 != null && t1 != null) "$t0 - $t1" else t0 ?: t1 ?: ""
        val dep = event.depositPercent ?: user?.defaultDepositPercent
        val ref = event.refundPercent ?: user?.defaultRefundPercent
        val can = event.cancellationDays ?: user?.defaultCancellationDays
        val svc = if (products.isNotEmpty())
            products.joinToString(", ") { "${it.quantity.toInt()} ${it.productName ?: "Producto"}" }
        else ""
        val city = event.city?.takeIf { it.isNotBlank() } ?: client.city?.takeIf { it.isNotBlank() } ?: ""
        val today = DATE_FMT.format(LocalDate.now())
        return mapOf(
            "Nombre del proveedor" to name,
            "Nombre comercial del proveedor" to biz,
            "Email del proveedor" to (user?.email ?: ""),
            "Fecha actual" to today,
            "Fecha del evento" to PdfConstants.formatDate(event.eventDate),
            "Hora de inicio" to (t0 ?: ""),
            "Hora de fin" to (t1 ?: ""),
            "Horario del evento" to range,
            "Tipo de servicio" to event.serviceType,
            "Número de personas" to event.numPeople.toString(),
            "Lugar del evento" to (event.location ?: ""),
            "Ciudad del evento" to (event.city ?: ""),
            "Servicios del evento" to svc,
            "Monto total del evento" to PdfConstants.formatCurrency(event.totalAmount),
            "Porcentaje de anticipo" to (dep?.toInt()?.toString() ?: ""),
            "Porcentaje de reembolso" to (ref?.toInt()?.toString() ?: ""),
            "Días de cancelación" to (can?.toInt()?.toString() ?: ""),
            "Total pagado" to "",
            "Nombre del cliente" to client.name,
            "Teléfono del cliente" to client.phone,
            "Email del cliente" to (client.email ?: ""),
            "Dirección del cliente" to (client.address ?: ""),
            "Ciudad del cliente" to (client.city ?: ""),
            "Ciudad del contrato" to city,
        )
    }

    private fun renderHardcoded(
        manager: PdfPageManager, event: Event, client: Client,
        products: List<EventProduct>, extras: List<EventExtra>, user: User?
    ) {
        manager.drawTitle("CONTRATO DE SERVICIOS")
        manager.drawText("Folio: ${event.id.take(8).uppercase()}", PdfConstants.secondaryTextPaint())
        manager.drawText("Fecha: ${PdfConstants.formatDate(event.createdAt.ifEmpty { event.eventDate })}", PdfConstants.secondaryTextPaint())
        manager.moveDown(PdfConstants.SECTION_SPACING)
        val bn = user?.businessName ?: "Solennix"
        manager.drawSectionHeader("PARTES")
        manager.drawMultilineText("El presente contrato se celebra entre $bn (en adelante \"EL PRESTADOR\") y ${client.name} (en adelante \"EL CLIENTE\").")
        manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
        manager.drawSectionHeader("DATOS DEL CLIENTE")
        manager.drawLabelValue("Nombre:", client.name)
        client.email?.let { manager.drawLabelValue("Email:", it) }
        manager.drawLabelValue("Telefono:", client.phone)
        client.address?.let { manager.drawLabelValue("Direccion:", it) }
        manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
        manager.drawSectionHeader("DETALLES DEL EVENTO")
        manager.drawLabelValue("Tipo de Servicio:", event.serviceType)
        manager.drawLabelValue("Fecha del Evento:", PdfConstants.formatDate(event.eventDate))
        event.startTime?.let { manager.drawLabelValue("Hora de Inicio:", it) }
        manager.drawLabelValue("Numero de Personas:", event.numPeople.toString())
        event.location?.let { manager.drawLabelValue("Ubicacion:", it) }
        event.city?.let { manager.drawLabelValue("Ciudad:", it) }
        manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
        if (products.isNotEmpty()) {
            manager.drawSectionHeader("SERVICIOS CONTRATADOS")
            products.forEach { p ->
                val total = p.totalPrice ?: (p.quantity * p.unitPrice)
                manager.drawText("- ${p.productName ?: "Producto"} x${p.quantity.toInt()} - ${PdfConstants.formatCurrency(total)}", PdfConstants.bodyPaint())
            }
            manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
        }
        if (extras.isNotEmpty()) {
            manager.drawSectionHeader("EXTRAS")
            extras.forEach { e -> manager.drawText("- ${e.description} - ${PdfConstants.formatCurrency(e.price)}", PdfConstants.bodyPaint()) }
            manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
        }
        manager.drawSectionHeader("CONDICIONES ECONOMICAS")
        val sub = products.sumOf { it.totalPrice ?: (it.quantity * it.unitPrice) } + extras.sumOf { it.price }
        manager.drawSummaryRow("Subtotal:", PdfConstants.formatCurrency(sub))
        if (event.discount > 0) {
            val da = if (event.discountType == DiscountType.PERCENT) sub * event.discount / 100 else event.discount
            val dt = if (event.discountType == DiscountType.PERCENT) "Descuento (${event.discount.toInt()}%):" else "Descuento:"
            manager.drawSummaryRow(dt, "-${PdfConstants.formatCurrency(da)}")
        }
        manager.moveDown(8f)
        manager.drawLine()
        manager.drawSummaryRow("TOTAL:", PdfConstants.formatCurrency(event.totalAmount), isBold = true)
        manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
        user?.defaultDepositPercent?.let { pct ->
            if (pct > 0) {
                val a = event.totalAmount * pct / 100
                manager.drawText("Anticipo requerido (${pct.toInt()}%): ${PdfConstants.formatCurrency(a)}", PdfConstants.boldPaint())
                manager.moveDown(PdfConstants.PARAGRAPH_SPACING)
            }
        }
        event.notes?.let { if (it.isNotBlank()) { manager.drawSectionHeader("NOTAS"); manager.drawMultilineText(it); manager.moveDown(PdfConstants.PARAGRAPH_SPACING) } }
        manager.moveDown(PdfConstants.SECTION_SPACING)
    }

    private fun drawHeader(manager: PdfPageManager, user: User?) {
        val bn = user?.businessName ?: "Solennix"
        manager.canvas.drawText(bn, PdfConstants.MARGIN_LEFT, manager.currentY, PdfConstants.titlePaint())
        manager.moveDown(PdfConstants.LINE_HEIGHT_NORMAL + 8f)
        manager.drawPrimaryLine()
        manager.moveDown(PdfConstants.SECTION_SPACING)
    }

    private fun drawSignatureSection(manager: PdfPageManager, businessName: String, clientName: String) {
        val sw = 200f
        val x1 = PdfConstants.MARGIN_LEFT + 30f
        val x2 = PdfConstants.PAGE_WIDTH - PdfConstants.MARGIN_RIGHT - sw - 30f
        val lp = android.graphics.Paint().apply { color = PdfConstants.COLOR_TEXT_SECONDARY; strokeWidth = 1f }
        manager.canvas.drawLine(x1, manager.currentY, x1 + sw, manager.currentY, lp)
        manager.canvas.drawLine(x2, manager.currentY, x2 + sw, manager.currentY, lp)
        manager.moveDown(PdfConstants.LINE_HEIGHT_SMALL)
        val lbl = PdfConstants.secondaryTextPaint().apply { textAlign = android.graphics.Paint.Align.CENTER }
        manager.canvas.drawText("EL PRESTADOR", x1 + sw / 2, manager.currentY, lbl)
        manager.canvas.drawText("EL CLIENTE", x2 + sw / 2, manager.currentY, lbl)
        manager.moveDown(PdfConstants.LINE_HEIGHT_NORMAL)
        val np = PdfConstants.bodyPaint().apply { textAlign = android.graphics.Paint.Align.CENTER }
        manager.canvas.drawText(businessName, x1 + sw / 2, manager.currentY, np)
        manager.canvas.drawText(clientName, x2 + sw / 2, manager.currentY, np)
    }
}

File: ios/Packages/SolennixFeatures/Sources/SolennixFeatures/Events/PDFGenerators/ContractPDFGenerator.swift
import UIKit
import SolennixCore

struct ContractPDFGenerator {

    /// Generates a contract PDF for the given event.
    /// Pass the full viewModel context (products, payments, productNames) so every
    /// token in the user's saved template can be resolved — mirror of the render
    /// logic in EventContractPreviewView.
    static func generate(
        event: Event,
        client: Client,
        profile: User?,
        products: [EventProduct] = [],
        payments: [Payment] = [],
        productNames: [String: String] = [:]
    ) -> Data {
        let renderer = UIGraphicsPDFRenderer(bounds: PDFConstants.pageRect)

        return renderer.pdfData { context in
            context.beginPage()

            // MARK: Header
            var y = PDFConstants.drawHeader(context: context, title: "Contrato de Servicios", profile: profile)

            // MARK: Contract Body
            let template = profile?.contractTemplate ?? defaultContractTemplate()
            let body = replaceTokens(
                in: template,
                event: event,
                client: client,
                profile: profile,
                products: products,
                payments: payments,
                productNames: productNames
            )

            y = drawContractBody(context: context, y: y, text: body)

            // MARK: Signature Boxes
            y += 24
            y = drawSignatureBoxes(context: context, y: y, profile: profile, client: client)

            // Page footer
            PDFConstants.drawFooterText(context: context, text: "Generado por Solennix")
        }
    }

    // MARK: - Token Replacement

    private static func replaceTokens(
        in template: String,
        event: Event,
        client: Client,
        profile: User?,
        products: [EventProduct],
        payments: [Payment],
        productNames: [String: String]
    ) -> String {
        let eventDateFormatted = PDFConstants.formatDate(event.eventDate)
        let todayFormatted = PDFConstants.dateFormatter.string(from: Date())
        let depositPercent = event.depositPercent ?? profile?.defaultDepositPercent ?? 0
        let depositAmount = event.totalAmount * (depositPercent / 100)
        let cancellationDays = event.cancellationDays ?? profile?.defaultCancellationDays ?? 0
        let refundPercent = event.refundPercent ?? profile?.defaultRefundPercent ?? 0
        let totalPaid = payments.reduce(0) { $0 + $1.amount }

        // Compute discount. Usamos el MISMO subtotal/formula que
        // EventFinancesDetailView para que el PDF matchee exacto los
        // numeros que el usuario ve en la UI. Evita la situacion donde
        // el cliente firma un contrato con un monto distinto al que se
        // le mostro en la app.
        let subtotalForDiscount = event.totalAmount - event.taxAmount
            + (event.discountType == .fixed ? event.discount : 0)
        let discountValue: Double = event.discountType == .percent
            ? subtotalForDiscount * event.discount / 100
            : event.discount

        // Compose complex tokens
        let scheduleValue: String = {
            if let s = event.startTime, let e = event.endTime { return "\(s) - \(e)" }
            return event.startTime ?? event.endTime ?? "—"
        }()

        let servicesValue: String = {
            guard !products.isEmpty else { return "—" }
            return products
                .map { "\(Int($0.quantity)) \(productNames[$0.productId] ?? "Producto")" }
                .joined(separator: ", ")
        }()

        let tokenMap: [String: String] = [
            "[Nombre del cliente]": client.name,
            "[Teléfono del cliente]": client.phone,
            "[Email del cliente]": client.email ?? "—",
            "[Dirección del cliente]": client.address ?? "—",
            "[Ciudad del cliente]": client.city ?? "—",
            "[Fecha del evento]": eventDateFormatted,
            "[Hora de inicio]": event.startTime ?? "—",
            "[Hora de fin]": event.endTime ?? "—",
            "[Horario del evento]": scheduleValue,
            "[Tipo de servicio]": event.serviceType,
            "[Número de personas]": "\(event.numPeople)",
            "[Ubicación del evento]": event.location ?? "—",
            "[Ciudad del evento]": event.city ?? "—",
            "[Lugar del evento]": event.location ?? "—",
            "[Monto total del evento]": PDFConstants.formatCurrency(event.totalAmount),
            "[Subtotal del evento]": PDFConstants.formatCurrency(event.totalAmount - event.taxAmount + discountValue),
            "[Descuento del evento]": PDFConstants.formatCurrency(discountValue),
            "[IVA del evento]": PDFConstants.formatCurrency(event.taxAmount),
            "[Porcentaje de anticipo]": "\(Int(depositPercent))%",
            "[Monto de anticipo]": PDFConstants.formatCurrency(depositAmount),
            "[Total pagado]": PDFConstants.formatCurrency(totalPaid),
            "[Días de cancelación]": "\(Int(cancellationDays))",
            "[Porcentaje de reembolso]": "\(Int(refundPercent))%",
            "[Nombre del negocio]": profile?.businessName ?? profile?.name ?? "—",
            "[Nombre comercial del proveedor]": profile?.businessName ?? profile?.name ?? "—",
            "[Nombre del proveedor]": profile?.name ?? "—",
            "[Email del proveedor]": profile?.email ?? "—",
            "[Fecha actual]": todayFormatted,
            "[Ciudad del contrato]": event.city ?? client.city ?? "—",
            "[Servicios del evento]": servicesValue,
            "[Notas del evento]": event.notes ?? ""
        ]

        var result = template
        for (token, value) in tokenMap {
            result = result.replacingOccurrences(of: token, with: value)
        }
        // Templates sometimes end up with literal '%%' when the author typed '%'
        // right after a percentage token (e.g. "[Porcentaje de anticipo]%"). The
        // token already renders "19%", so the adjacent '%' produces "19%%".
        // Normalize any resulting double-percent back to single.
        result = result.replacingOccurrences(of: "%%", with: "%")
        return result
    }

    // MARK: - Drawing Helpers

    private static func drawContractBody(context: UIGraphicsPDFRendererContext, y startY: CGFloat, text: String) -> CGFloat {
        let attrs: [NSAttributedString.Key: Any] = [
            .font: PDFConstants.bodyFont,
            .foregroundColor: PDFConstants.textColor
        ]

        let paragraphs = text.components(separatedBy: "\n")
        var currentY = startY

        for paragraph in paragraphs {
            if paragraph.trimmingCharacters(in: .whitespaces).isEmpty {
                currentY += 8
                continue
            }

            let str = paragraph as NSString
            let size = str.boundingRect(
                with: CGSize(width: PDFConstants.contentWidth, height: .greatestFiniteMagnitude),
                options: .usesLineFragmentOrigin,
                attributes: attrs,
                context: nil
            )

            // Check if we need a new page
            if currentY + size.height + 4 > PDFConstants.pageHeight - PDFConstants.marginBottom {
                PDFConstants.drawFooterText(context: context, text: "Generado por Solennix")
                context.beginPage()
                currentY = PDFConstants.marginTop
            }

            str.draw(
                in: CGRect(x: PDFConstants.marginLeft, y: currentY, width: PDFConstants.contentWidth, height: size.height + 2),
                withAttributes: attrs
            )
            currentY += size.height + 4
        }

        return currentY
    }

    private static func drawSignatureBoxes(context: UIGraphicsPDFRendererContext, y startY: CGFloat, profile: User?, client: Client) -> CGFloat {
        let boxHeight: CGFloat = 80
        var y = PDFConstants.ensureSpace(context: context, currentY: startY, needed: boxHeight + 40)

        let colWidth = PDFConstants.contentWidth / 2 - 10
        let leftX = PDFConstants.marginLeft
        let rightX = PDFConstants.marginLeft + PDFConstants.contentWidth / 2 + 10

        let labelAttrs: [NSAttributedString.Key: Any] = [
            .font: PDFConstants.bodyBoldFont,
            .foregroundColor: PDFConstants.textColor
        ]
        let nameAttrs: [NSAttributedString.Key: Any] = [
            .font: PDFConstants.bodyFont,
            .foregroundColor: PDFConstants.grayColor
        ]

        // Left: EL PROVEEDOR
        ("EL PROVEEDOR" as NSString).draw(
            in: CGRect(x: leftX, y: y, width: colWidth, height: 16),
            withAttributes: labelAttrs
        )

        // Signature line
        let lineY = y + 50
        let linePath = UIBezierPath()
        linePath.move(to: CGPoint(x: leftX, y: lineY))
        linePath.addLine(to: CGPoint(x: leftX + colWidth, y: lineY))
        linePath.lineWidth = 0.5
        PDFConstants.grayColor.setStroke()
        linePath.stroke()

        let providerName = profile?.name ?? "—"
        (providerName as NSString).draw(
            in: CGRect(x: leftX, y: lineY + 4, width: colWidth, height: 16),
            withAttributes: nameAttrs
        )

        // Right: EL CLIENTE
        ("EL CLIENTE" as NSString).draw(
            in: CGRect(x: rightX, y: y, width: colWidth, height: 16),
            withAttributes: labelAttrs
        )

        let linePath2 = UIBezierPath()
        linePath2.move(to: CGPoint(x: rightX, y: lineY))
        linePath2.addLine(to: CGPoint(x: rightX + colWidth, y: lineY))
        linePath2.lineWidth = 0.5
        PDFConstants.grayColor.setStroke()
        linePath2.stroke()

        (client.name as NSString).draw(
            in: CGRect(x: rightX, y: lineY + 4, width: colWidth, height: 16),
            withAttributes: nameAttrs
        )

        return lineY + 24
    }

    // MARK: - Default Template

    private static func defaultContractTemplate() -> String {
        """
        CONTRATO DE PRESTACIÓN DE SERVICIOS

        En la ciudad de [Ciudad del evento], a [Fecha actual], comparecen por una parte [Nombre del negocio], en lo sucesivo "EL PROVEEDOR", y por otra parte [Nombre del cliente], en lo sucesivo "EL CLIENTE".

        DECLARACIONES

        EL PROVEEDOR declara que cuenta con la capacidad y experiencia para proporcionar servicios de [Tipo de servicio].

        EL CLIENTE declara que requiere los servicios de EL PROVEEDOR para el evento a celebrarse el día [Fecha del evento] en [Ubicación del evento].

        CLÁUSULAS

        PRIMERA. OBJETO DEL CONTRATO
        EL PROVEEDOR se compromete a prestar servicios de [Tipo de servicio] para [Número de personas] personas el día [Fecha del evento], con horario de [Hora de inicio] a [Hora de fin].

        SEGUNDA. PRECIO Y FORMA DE PAGO
        El precio total de los servicios será de [Monto total del evento]. EL CLIENTE deberá cubrir un anticipo del [Porcentaje de anticipo] ([Monto de anticipo]) al momento de la firma del presente contrato. El saldo restante deberá cubrirse a más tardar el día del evento.

        TERCERA. CANCELACIÓN
        En caso de cancelación por parte de EL CLIENTE con menos de [Días de cancelación] días de anticipación, EL PROVEEDOR reembolsará el [Porcentaje de reembolso] del anticipo.

        CUARTA. OBLIGACIONES DEL PROVEEDOR
        EL PROVEEDOR se obliga a proporcionar los servicios pactados en tiempo y forma, conforme a las especificaciones acordadas.

        QUINTA. OBLIGACIONES DEL CLIENTE
        EL CLIENTE se obliga a realizar los pagos en los plazos acordados y a proporcionar las facilidades necesarias para la prestación del servicio.

        Leído el presente contrato, ambas partes lo firman de conformidad.
        """
    }
}
